### PR TITLE
add rich message component support to configuration

### LINKED ISSUE #5449: Add Component-Methods to FileConfigurations
### Is your feature request related to a problem?

The problem is that with the new Component-System, Adventure, that has been put into the Paper-API, everything's pretty much fine, but FileConfigurations, like the default config you can add to your plugin, does currently not have any methods related to Components.

### Describe the solution you'd like.

It would be great to have methods for FileConfigurations regarding Components.
Such as:
getComponent(String path), which returns a Component,
getComponentList(String path) , which returns a List consisting of Components,
and maybe also getText and get TextList, for TextComponents, but that one is just QoL.

Because getString and getStringList both exist, it would be great to have the same methods for Components, without having to convert to one from a string or List<String>.

### Describe alternatives you've considered.

It is possible to get a plain String out of the config and then convert that one to a Component or TextComponent, 
and it is also possible, easiest with an ArrayStream, to convert a List<String> to a List<Component>, but if Adventure is now the only way to go, and all the methods containing Strings are deprecated now, it would be great to have simple methods for the new Types.

### Agreements

I have searched for and ensured there isn't already an open issue regarding this.
I have ensured the feature I'm requesting isn't already in the latest supported Paper build.

### PR DESCRIPTION
**Pull request by yannicklamprecht**
resolves #5449 

TODOS:

- [x] Wait for adventure 4.16.0 release
- [x] remove mavenLocal() and migrate to version 4.16.0
- [x] add changes to adventure patch

### CODE REVIEW

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
We should probably allow the ability to pass custom MiniMessage instances for this for custom tags and whatnot.

**Comment by imDaniX** [comment]
> We should probably allow the ability to pass custom MiniMessage instances for this for custom tags and whatnot.

Having an option to pass custom `ComponentSerializer` in general would be helpful too.

As a side note, this will currently throw an exception if a user uses `§` in their string. I'm just thinking that there should be some safe railings for this case (like returning a def/null component), or at least some javadoc info mentioning the behavior.

**Comment by yannicklamprecht** [comment]
> We should probably allow the ability to pass custom MiniMessage instances for this for custom tags and whatnot.

Done.

**Review by lynxplay — DISMISSED** [review]
I think it would be great if we split this into 

getRichMessage(path)
getRichMessage(path, defaultMini)

getComponent(path, ComponentSerialiser<T, String>, defaultStr)

that way, we also enable easy fetching for alternative formats down the line.

File: patches/api/0459-add-rich-message-component-support-to-configuration.patch [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
++     * this will return the default value. If the Component does not exist and no
++     * default value was specified, this will return null.
++     *
++     * @param path Path of the Component to get.
++     * @param component component that will be used as fallback
++     * @return Requested Component.
++     */
++    @Contract("_, !null -> !null")
++    @Nullable
++    default net.kyori.adventure.text.Component getRichMessage(@NotNull String path, @Nullable net.kyori.adventure.text.Component component){
```
Comment (yannicklamprecht) [inline patches/api/0459-add-rich-message-component-support-to-configuration.patch]: @lynxplay should I yeet this?

Comment (yannicklamprecht) [reply patches/api/0459-add-rich-message-component-support-to-configuration.patch]: As discussed in voice, we instead yank the String default one.

> We have String Templates in one Java version anyways. So we just have a MiniMessage StringProcessor.

File: patches/api/0460-add-rich-message-component-support-to-configuration.patch [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
++     * default value was specified, this will return null.
++     *
++     * @param path Path of the Component to get.
++     * @param componentSerializer componentSerializer instance used for deserialization
++     * @param component component that will be used as fallback
++     * @return Requested Component.
++     */
++    @Contract("_, _, !null -> !null")
++    @Nullable
++    net.kyori.adventure.text.Component getComponent(@NotNull String path, @NotNull net.kyori.adventure.text.serializer.ComponentSerializer<net.kyori.adventure.text.Component, net.kyori.adventure.text.Component, String> componentSerializer, @Nullable net.kyori.adventure.text.Component component);
```
Comment (Machine-Maker) [inline patches/api/0460-add-rich-message-component-support-to-configuration.patch]: This method should use generics to support serializers that output other Component subtypes

prob smth like `<C extends Component> C getComponent(..., ComponentSerializer<C, ?, String> serializer, ...)`

**Review by kennytv — APPROVED** [review]

File: patches/api/0460-add-rich-message-component-support-to-configuration.patch [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
++    /**
++     * Gets the requested Component by path.
++     * <p>
++     * If the Component does not exist but a default value has been specified,
++     * this will return the default value. If the Component does not exist and no
++     * default value was specified, this will return null.
++     *
++     * @param path Path of the Component to get.
++     * @return Requested Component.
++     */
```
Comment (emilyy-dev) [inline patches/api/0460-add-rich-message-component-support-to-configuration.patch]: I think it's worth noting that the component at the given path is a MiniMessage string, the same for the (potential) default value and the other getRichMessage method

File: patches/api/0460-add-rich-message-component-support-to-configuration.patch [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
++     * this will return the default value. If the Component does not exist and no
++     * default value was specified, this will return null.
++     *
++     * @param path Path of the Component to get.
++     * @param component component that will be used as fallback
++     * @return Requested Component.
++     */
++    @Contract("_, !null -> !null")
++    @Nullable
++    default net.kyori.adventure.text.Component getRichMessage(@NotNull String path, @Nullable net.kyori.adventure.text.Component component){
```
Comment (emilyy-dev) [inline patches/api/0460-add-rich-message-component-support-to-configuration.patch]: rename `component` parameter to `fallback`/`def`, same with the `getComponent` method

File: patches/api/0460-add-rich-message-component-support-to-configuration.patch [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
++     * default value was specified, this will return null.
++     *
++     * @param path Path of the Component to get.
++     * @param componentSerializer componentSerializer instance used for deserialization
++     * @param component component that will be used as fallback
++     * @return Requested Component.
++     */
++    @Contract("_, _, !null -> !null")
++    @Nullable
++    <C extends net.kyori.adventure.text.Component> C getComponent(@NotNull String path, @NotNull net.kyori.adventure.text.serializer.ComponentSerializer<C, C, String> componentSerializer, @Nullable C component);
```
Comment (emilyy-dev) [inline patches/api/0460-add-rich-message-component-support-to-configuration.patch]: Do we care about a fallback-lacking overload? `getComponent(String, ComponentSerializer)`

File: patches/api/0460-add-rich-message-component-support-to-configuration.patch [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+     }
+ 
++    // Paper start - add rich message component support to configuration
++    @Nullable
++    @Override
++    public <C extends net.kyori.adventure.text.Component> C getComponent(@NotNull final String path, @NotNull final net.kyori.adventure.text.serializer.ComponentSerializer<C, C, String> componentSerializer, @Nullable final C component) {
++        final String value = getString(path);
++        return value == null ? component : componentSerializer.deserialize(value);
++    }
++    // Paper end - add rich message component support to configuration
```
Comment (emilyy-dev) [inline patches/api/0460-add-rich-message-component-support-to-configuration.patch]: this can be promoted to `ConfigurationSection` as a `default` method

**Review by emilyy-dev — COMMENTED** [review]
Generally looks good, my couple concerns are regarding documentation mostly

**Comment by lynxplay** [comment]
General "lets wait and see" for https://github.com/KyoriPowered/adventure/issues/1034, as we could use ComponentDecoder over ComponentSerialiser if possible.

I'd give it a couple days, otherwise we'll commodore it down the line.

**Comment by emilyy-dev** [comment]
Also maybe the patch could/should be promoted to the adventure patch, would like to know if anyone has strong opinions against that

**Comment by yannicklamprecht** [comment]
> General "lets wait and see" for [KyoriPowered/adventure#1034](https://github.com/KyoriPowered/adventure/issues/1034), as we could use ComponentDecoder over ComponentSerialiser if possible.
> 
> I'd give it a couple days, otherwise we'll commodore it down the line.

Waiting another week or month will not kill it. So just wait for the decoder.

File: patches/api/0460-add-rich-message-component-support-to-configuration.patch (hunk 3 above)
Comment (yannicklamprecht) [reply patches/api/0460-add-rich-message-component-support-to-configuration.patch]: Hopefully addressed like you expected it.

File: patches/api/0461-add-rich-message-component-support-to-configuration.patch [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
++     * @return Requested Component.
++     */
++    @Contract("_, !null -> !null")
++    @Nullable
++    default net.kyori.adventure.text.Component getRichMessage(@NotNull String path, @Nullable net.kyori.adventure.text.Component fallback) {
++        return getComponent(path, net.kyori.adventure.text.minimessage.MiniMessage.miniMessage(), fallback);
++    }
++
++    /**
++     * Gets the requested MiniMessage formatted String as Component by path.
```
Comment (Machine-Maker) [inline patches/api/0461-add-rich-message-component-support-to-configuration.patch]: Not always `MiniMessage` now that it takes any serializer

File: patches/api/0461-add-rich-message-component-support-to-configuration.patch [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
++     * default value was specified, this will return null.
++     *
++     * @param path Path of the Component to get.
++     * @param componentSerializer componentSerializer instance used for deserialization
++     * @param fallback component that will be used as fallback
++     * @return Requested Component.
++     */
++    @Contract("_, _, !null -> !null")
++    @Nullable
++    default <C extends net.kyori.adventure.text.Component> C getComponent(@NotNull String path, @NotNull net.kyori.adventure.text.serializer.ComponentSerializer<C, C, String> componentSerializer, @Nullable C fallback) {
```
Comment (Machine-Maker) [inline patches/api/0461-add-rich-message-component-support-to-configuration.patch]: I think the first generic param should be `?` cause we technically don't care at all about that input type, we only care about the output type.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
Should we add `set` methods as well? I don't think there are for other specific types, but since we aren't hooking Component into the snakeyaml type system, maybe some set methods would be good?

File: patches/api/0461-add-rich-message-component-support-to-configuration.patch (hunk 8 above)
Comment (emilyy-dev) [reply patches/api/0461-add-rich-message-component-support-to-configuration.patch]: > I think the first generic param should be `?` cause we technically don't care at all about that input type, we only care about the output type.

Although if we wait for ComponentDecoder to be a thing this wouldn't be a concern

**Comment by yannicklamprecht** [comment]
> Should we add `set` methods as well? I don't think there are for other specific types, but since we aren't hooking Component into the snakeyaml type system, maybe some set methods would be good?

Addressed with 72031ae5a6b410c505cd0285a1003897d1b31449

File: patches/api/0461-add-rich-message-component-support-to-configuration.patch [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
++     *
++     * @param path Path of the Component to get.
++     * @param decoder ComponentDecoder instance used for deserialization
++     * @param fallback component that will be used as fallback
++     * @return Requested Component.
++     */
++    @Contract("_, _, !null -> !null")
++    @Nullable
++    default <C extends net.kyori.adventure.text.Component> C getComponent(@NotNull String path, @NotNull net.kyori.adventure.text.serializer.ComponentSerializer<?, C, String> decoder, @Nullable C fallback) {
++        com.google.common.base.Preconditions.checkArgument(decoder != null, "Decoder needs to be present");
```
Comment (kashike) [inline patches/api/0461-add-rich-message-component-support-to-configuration.patch]:
```suggestion
+        java.util.Objects.requireNonNull(decoder, "decoder");
```

File: patches/api/0461-add-rich-message-component-support-to-configuration.patch [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
++     * <p>
++     * If value is null, the entry will be removed. Any existing entry will be
++     * replaced, regardless of what the new value is.
++     *
++     * @param path Path of the object to set.
++     * @param encoder the encoder used to transform the value
++     * @param value New value to set the path to.
++     */
++    default <I extends net.kyori.adventure.text.Component> void setComponent(@NotNull String path, @NotNull net.kyori.adventure.text.serializer.ComponentEncoder<I, String> encoder, @Nullable I value) {
++        com.google.common.base.Preconditions.checkArgument(encoder != null, "Encoder needs to be present");
```
Comment (kashike) [inline patches/api/0461-add-rich-message-component-support-to-configuration.patch]:
```suggestion
+        java.util.Objects.requireNonNull(encoder, "encoder");
```

File: patches/api/0461-add-rich-message-component-support-to-configuration.patch [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+diff --git a/src/main/java/org/bukkit/configuration/MemorySection.java b/src/main/java/org/bukkit/configuration/MemorySection.java
+index 90751edd742d9b8d2171c4b16891aa009e4b8dfe..e057b03671f091eccc47ae9961b0ceb37d40504a 100644
+--- a/src/main/java/org/bukkit/configuration/MemorySection.java
++++ b/src/main/java/org/bukkit/configuration/MemorySection.java
+@@ -189,6 +189,12 @@ public class MemorySection implements ConfigurationSection {
+     @Override
+     public void set(@NotNull String path, @Nullable Object value) {
+         Preconditions.checkArgument(!Strings.isNullOrEmpty(path), "Cannot set to an empty path");
++        // Paper start - add rich message component support to configuration
++        if(value instanceof net.kyori.adventure.text.Component component) {
```
Comment (kashike) [inline patches/api/0461-add-rich-message-component-support-to-configuration.patch]:
```suggestion
+        if (value instanceof final net.kyori.adventure.text.Component component) {
```

**Review by kashike — DISMISSED** [review]

File: patches/api/0460-add-rich-message-component-support-to-configuration.patch (hunk 5 above)
Comment (yannicklamprecht) [reply patches/api/0460-add-rich-message-component-support-to-configuration.patch]: Addressed in https://github.com/PaperMC/Paper/commit/88393c02489850bbf8d582bf0da56bdea82fd408#diff-641780de4795f475f2b5c979f77ab04d7be650885ee99dd34b574ea6a0209983R74-R76

File: patches/api/0461-add-rich-message-component-support-to-configuration.patch (hunk 8 above)
Comment (yannicklamprecht) [reply patches/api/0461-add-rich-message-component-support-to-configuration.patch]: Addressed by using a local version of https://github.com/KyoriPowered/adventure/pull/1035 see https://github.com/PaperMC/Paper/pull/10225/commits/45b2e17f141e167b861f15b21dc9c85277fb5786

**Comment by yannicklamprecht** [comment]
Should we fixup this onto the adventure commit?

**Comment by lynxplay** [comment]
Yea :+1: add yourself as coauthor but yea

**Comment by yannicklamprecht** [comment]
> Yea 👍 add yourself as coauthor but yea

Done.

**Review by electronicboy — APPROVED** [review]

**Review by emilyy-dev — APPROVED** [review]

**Review by Owen1212055 — APPROVED** [review]

**Review by Machine-Maker — APPROVED** [review]

**Pull request merged**